### PR TITLE
feat: `preview_persisted_queries` w/opt-in safelisting

### PR DESCRIPTION
**Pull request by EverlastingBugstopper**
# Persisted Queries w/opt-in safelisting

> ⚠️ **This is an [Enterprise feature](https://www.apollographql.com/blog/platform/evaluating-apollo-router-understanding-free-and-open-vs-commercial-features/) of the Apollo Router.** It requires an organization with a [GraphOS Enterprise plan](https://www.apollographql.com/pricing/) and the feature to be enabled for your account.
>
> If your organization _doesn't_ currently have an Enterprise plan, you can test out this functionality by signing up for a free [Enterprise trial](https://www.apollographql.com/docs/graphos/org/plans/#enterprise-trials) and reaching out to enable the feature for your account.

## Overview

The persisted queries feature allows you to pre-register operations, allowing clients to send an operation ID over the wire and execute the associated operation. Each operation defines the exact shape of a GraphQL operation that the router expects clients to send. In its simplest form, Persisted Queries (PQ’s) can be used like Automatic Persisted Queries (APQ’s) with one key difference: sending an operation body is never allowed for a PQ. Registering persisted operations allows locking down the router to log unregistered operations, or to reject them outright.

### Main Configurations

* **Unregistered operation monitoring**
  * Your router can allow all GraphQL operations, while emitting structured traces containing unregistered operation bodies.
* **Operation safelisting** 
  * Reject unregistered operations
  * Require all operations to be sent as an ID

## Usage

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
```

This enables additive PQs.

Requires `APOLLO_KEY` and `APOLLO_GRAPH_REF` to start up properly (to fetch the license key and the persisted queries themselves), and the graph variant must be linked to a persisted query list. This is only available in preview right now and has to be enabled for a graph.

The router will not start up until all persisted queries have been read into a `std::collections::HashMap<String, String>` mapping ID to their body. Additionally, just the bodies are stored in a `std::collections::HashSet`.

After the router starts, persisted queries can be sent over the wire like so:

```sh
curl http://localhost:4000/ -X POST --json \
'{"extensions":{"persistedQuery":{"version":1,"sha256Hash":"dc67510fb4289672bea757e862d6b00e83db5d3cbbcfb15260601b6f29bb2b8f"}}}'
```

2) [./examples/persisted-queries/safelist_pq_log_only.yaml](https://github.com/apollographql/router/raw/avery/persisted-queries/examples/persisted-queries/safelist_pq_log_only.yaml)

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
  log_unpersisted_queries: true
```

Starting the router with this configuration logs freeform GraphQL operations that do not match a persisted query.

3) [./examples/persisted-queries/safelist_pq.yaml](https://github.com/apollographql/router/raw/avery/persisted-queries/examples/persisted-queries/safelist_pq.yaml)

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
  safelist:
    enabled: true
apq:
  enabled: false
```

Starting the router with this configuration will require all operations sent over the wire to match either the ID (O(1) retrieval from `HashMap`) or the body (O(1) retrieval from `HashSet`). APQ is enabled by default, and is incompatible with the persisted queries feature (clients are not allowed to register their own persisted queries, they must be pre-published), therefore it must be disabled to start properly. An error is returned if APQ is not explicitly disabled in `router.yaml`. 

4) [./examples/persisted-queries/safelist_pq_require_id.yaml](https://github.com/apollographql/router/raw/avery/persisted-queries/examples/persisted-queries/safelist_pq_require_id.yaml)

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
  safelist:
    enabled: true
    require_id: true
apq:
  enabled: false
```

This configuration is a stricter version of safelisting that rejects all freeform GraphQL requests, even if they match the body of a persisted query.

### CODE REVIEW

File: apollo-router/src/lib.rs
Comment (o0Ignition0o) [inline apollo-router/src/lib.rs:74]: is this intentional? iirc we d rather pub use (like at line 92 for example)

File: apollo-router/Cargo.toml
Comment (o0Ignition0o) [inline apollo-router/Cargo.toml:218]: Should it be a dev-dependency?

File: apollo-router/src/services/layers/persisted_queries/manifest_poller.rs
Comment (o0Ignition0o) [inline apollo-router/src/services/layers/persisted_queries/manifest_poller.rs:175]: it's in a separate thread, so there might be a way to recover from that 🤔

File: apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new
Comment (o0Ignition0o) [inline apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new:13]: I don't think this is what is expected.
This often means the query planner was updated and the mock was used to work with `{errorField id}`

⚠️ i think this needs to be addressed before merging

File: apollo-router/src/services/layers/persisted_queries/mod.rs
Comment (o0Ignition0o) [inline apollo-router/src/services/layers/persisted_queries/mod.rs:156]: I don't understand why there's an assert here, i'd rather have us not do that, and possibly adding an integration test that ensures a router doesn't start up if both are enabled?

**Review by o0Ignition0o — APPROVED** [review]
Preapproving, I still have two questions, one around an assert! and the other one about the snapshot.

Great work! 🚀

File: apollo-router/Cargo.toml
Comment (o0Ignition0o) [reply apollo-router/Cargo.toml:218]: fixed since it will be followed up on

File: apollo-router/src/services/layers/persisted_queries/manifest_poller.rs
Comment (EverlastingBugstopper) [reply apollo-router/src/services/layers/persisted_queries/manifest_poller.rs:175]: i'd love to see how this would work! no clue how i'd do that 😆

File: apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new
Comment (EverlastingBugstopper) [reply apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new:13]: agree, not sure how this made it in - thought i fixed that up.

File: apollo-router/src/services/layers/persisted_queries/mod.rs
Comment (EverlastingBugstopper) [reply apollo-router/src/services/layers/persisted_queries/mod.rs:156]: yeah we do have that test, i added this on a suggestion from @glasser.

Comment (glasser) [reply apollo-router/src/services/layers/persisted_queries/mod.rs:156]: It definitely seems pretty vital to ensure that later refactorings to how various features interact with each other don't lead to create a giant security hole in the safelist. I suppose just having an integration test might be good enough, but this logic is a bit hairy and the idea of having a codepath that says "I see a body was sent and I'm going to allow it" that doesn't allude to the concept of safelisting seemed scary. Perhaps the entire `supergraph_request` function could be refactored so that the structure is more like

```
if (there is a body) {
  ... do something
} else if (there is an ID) {
  ... do something
}
```

instead of having the top-level logic be "if there is an ID" — that might be a structure that would naturally avoid this assertion.  but in the current structure it really does feel to me that we want to be really really sure that there's no safelist before we say "this document, which we haven't checked against the PQ list, is legit".  (a previous version of this PR did have a bug here where you could effectively circumvent the safelist merely by including a random ID along with a document!)

Comment (glasser) [reply apollo-router/src/services/layers/persisted_queries/mod.rs:156]: (and I'm not necessarily recommending that restructure as something worth doing)

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new
Comment (BrynCooke) [reply apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new:13]: If possible can we make the test panic if it detects a missing mock?

File: apollo-router/src/lib.rs
Comment (EverlastingBugstopper) [reply apollo-router/src/lib.rs:74]: I import the mocks with `use crate::test_harness::mocks::persisted_queries::*;` and kind of like that structure? I don't want to expose individual mock structs for something specific to the PQ layer.

File: apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new
Comment (EverlastingBugstopper) [reply apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new:13]: what do you mean?

Comment (BrynCooke) [reply apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__errors_from_primary_on_deferred_responses.snap.new]: If the snapshot contains `couldn't find mock for query` it should hard fail the test.

File: apollo-router/src/services/layers/persisted_queries/manifest_poller.rs
Comment (BrynCooke) [reply apollo-router/src/services/layers/persisted_queries/manifest_poller.rs:175]: Usually for locks we use `expect` as recovery from poisoning is is very difficult to do safely and introduces unexpected lifecycle handling. i.e. you have to throw everything away. Honestly I think that we should never try to recover from poisoned locks.

**Comment by BrynCooke** [comment]
I'm going to fix the remaining test failures, make some minor modifications and then merge this.

File: apollo-router/src/services/layers/persisted_queries/mod.rs
Comment (BrynCooke) [reply apollo-router/src/services/layers/persisted_queries/mod.rs:156]: I'm going to raise this as a follow up ticket. We're still in preview, and we have time to change this.

File: apollo-router/src/services/layers/persisted_queries/manifest_poller.rs
Comment (o0Ignition0o) [reply apollo-router/src/services/layers/persisted_queries/manifest_poller.rs:175]: all right, wfm!

File: apollo-router/src/services/layers/persisted_queries/mod.rs
Comment (o0Ignition0o) [reply apollo-router/src/services/layers/persisted_queries/mod.rs:156]: Followup sounds good, I'm extremely uncomfortable with asserts in production code (and panics / expects for that matter). we can failfast_debug! and / or return an error, but the router should never ever crash unless it's at startup (it shouldnt crash at reload for example).

I understand this codepath is `unreachable!`, but better safe than sorry.

Comment (glasser) [reply apollo-router/src/services/layers/persisted_queries/mod.rs:156]: I don't have any strong feelings about *how* returning `Ok(request)` in this case is prevented — returning a 5xx to the user in this hopefully-unreachable case would be fine too. (But also if this didn't block merge/alpha that would be great :) )

**Comment by o0Ignition0o** [comment]
I'm going to bypass branch protections and merge this so we can cut an alpha release today, for testing purposes.

**Comment by glasser** [comment]
(The only branch protection bypassed was the Performance Tests.)

**Pull request merged**
